File: python/bin/p003.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Project Euler problem 3"""

def divide_out_factor(x, factor):
    while x % factor == 0:
        x = x // factor
    return x

def largest_prime_factor(x):
    x = divide_out_factor(x, 2)
    if x == 1:
        return 2
    x = divide_out_factor(x, 3)
    if x == 1:
        return 3
    y = 5
    while True:
        x = divide_out_factor(x, y)
        if x == 1:
            return y
        y += 2
        x = divide_out_factor(x, y)
        if x == 1:
            return y
        y += 4

def main():
    p = largest_prime_factor(600851475143)
    print(p)

if __name__ == "__main__":
    main()

File: c/src/p003.c
#include <stdio.h>
#include <inttypes.h>

#define true (1)

int64_t divide_out_factor(int64_t x, int64_t factor) {
    while (x % factor == 0) {
        x /= factor;
    }
    return x;
}

int64_t largest_prime_factor(int64_t x) {
    int64_t y;

    x = divide_out_factor(x, 2);
    if (x == 1) return 2;

    x = divide_out_factor(x, 3);
    if (x == 1) return 3;
    y = 5;
    while (true) {
        x = divide_out_factor(x, y);
        if (x == 1) return y;
        y += 2;

        x = divide_out_factor(x, y);
        if (x == 1) return y;
        y += 4;
    }
}

int main(void) {
    int64_t p = largest_prime_factor(600851475143);
    printf("%d\n", (int)p);
    return 0;
}
